SSA/Projects/Dialect.lean

```diff
@@ -0,0 +1,78 @@
+import SSA.Projects.InstCombine.InstCombineBase
+import SSA.Projects.InstCombine.Tests
+
+inductive Dialect
+  | InstCombine : Dialect
+  deriving Repr
+
```

Comment:
`Dialect` looks like it would be very suited as a typeclass (at some point in the future, might be premature generalization right now)

Reply:
I believe it should be an `AttribtuteEnvironmentExtension` so it is in fact open world. Users can register dialects by tagging a `structure Dialect ` with `@[dialect]`.

Reply:
That's a fair point. My reason for suggesting the typeclass is that the current approach seems like it will become unmanageable when there are more dialects, since the different operations that need to be defined for that dialect become quite spread out, and operation `Dialect.foo` has to be defined for all dialect in one place. Maybe a middle road is to group the functions that define a `Dialect` into a single structure 
```lean
structure DialectInterface where
  baseType : Dialect → Type
  userType : Dialect → Type
  opType : Dialect → Type
  ..
```

And then have a single function `Dialect.toInterface : Dialect -> DialectInterface`. 

Although, again, maybe a concern for later when we actually need more than just the one dialect

Reply:
Yep, I agree with all of that @alexkeizer! 

I just think it's two separate issues. I think managing all of those is the right thing for a typeclass (and I would use that to do something like what `TUSInstance` is doing in #37 )! The problem is that we need some way of keeping track of the different dialects we have and use the right types/instances for those depending on the dialect, like we do now in `Main.lean`. That's a separate concern.

Maybe I can do a consolidating PR once this and #37 are merged and deal with it there?

Reply:
I certainly wouldn't say this should block the current PR, so dealing with it in a consolidating PR seems fine!